### PR TITLE
Added JAX domain example, vision image segmentation

### PR DESCRIPTION
**Pull request by vfdev-5**

### CODE REVIEW

**Comment by vfdev-5** [comment]
@melissawm can you please review

File: docs/JAX_examples_image_segmentation.md [hunk 1]
```diff
@@ -0,0 +1,1389 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.15.2
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+## Image segmentation with JAX
```
Comment (jakevdp) [inline docs/JAX_examples_image_segmentation.md:15]: This should be an H1 header

**Comment by jakevdp** [comment]
One last build issue:
```
/home/docs/checkouts/readthedocs.org/user_builds/jax-ai-stack/checkouts/93/docs/JAX_examples_image_segmentation.ipynb:290006: WARNING: image file not readable: attachment:ed48cb36-b8c8-442f-b626-28e403b61ca2.png [image.not_readable]
```

**Comment by vfdev-5** [comment]
There is a warning in the readthedocs:
```
/home/docs/checkouts/readthedocs.org/user_builds/jax-ai-stack/checkouts/93/docs/JAX_examples_image_segmentation.ipynb:290006: WARNING: image file not readable: attachment:ed48cb36-b8c8-442f-b626-28e403b61ca2.png [image.not_readable]
looking for now-outdated files... none found
```
complaining about this text : 
```
"![image.png](attachment:ed48cb36-b8c8-442f-b626-28e403b61ca2.png)\n",
```
Image is attached in the jupyter notebook:
```
   "attachments": {
    "ed48cb36-b8c8-442f-b626-28e403b61ca2.png": {
     "image/png": <data>
    }
```
but this seems like not supported in readthedocs...
Should we create `assets` folder in the docs?

File: docs/JAX_examples_image_segmentation.md [hunk 2]
```diff
@@ -0,0 +1,1389 @@
...
+for img, mask in zip(images[:3], masks[:3]):
+    display_datapoint({"image": img, "mask": mask}, label=" (augmented validation set)")
+```
+
+## Model for Image Segmentation
+
+In this section we will implement from scratch [UNETR](https://arxiv.org/abs/2103.10504) model using Flax NNX. Reference PyTorch implementation of this model can be found on the GitHub in [MONAI library](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/nets/unetr.py).
+
+The UNETR model utilizes a transformer as the encoder to learn sequence representations of the input and to capture the global multi-scale information, while also following the “U-shaped” network design like [UNet](https://arxiv.org/abs/1505.04597) model:
+![image.png](attachment:ed48cb36-b8c8-442f-b626-28e403b61ca2.png)
```
Comment (jakevdp) [inline docs/JAX_examples_image_segmentation.md:351]: Did you mean to include this image in the PR?

Comment (jakevdp) [reply docs/JAX_examples_image_segmentation.md:351]: It looks like notebook attachments are not supported: https://github.com/executablebooks/MyST-NB/issues/230

Maybe we can put it separately in an `html_static_path`

File: docs/JAX_examples_image_segmentation.md [hunk 3]
```diff
@@ -0,0 +1,1387 @@
...
+    xentropy_loss = optax.softmax_cross_entropy_with_integer_labels(
+        logits=logits, labels=masks
+    ).mean()
+
+    jacc_loss = compute_softmax_jaccard_loss(logits=logits, masks=masks)
+    loss = xentropy_loss + jacc_loss
+    return loss, (xentropy_loss, jacc_loss, logits)
+```
+
+Let us implement confusion matrix metric derived from [`nnx.Metric`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/metrics.html#flax.nnx.metrics.Metric). Confusion matrix will help us to compute Intersection-Over-Union (IoU) metric: per class and the average. Finally, we can also compute accuracy metric using the confusion matrix.
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1079]:
```suggestion
Now, we will implement a confusion matrix metric derived from [`nnx.Metric`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/training/metrics.html#flax.nnx.metrics.Metric). A confusion matrix will help us to compute the Intersection-Over-Union (IoU) metric per class and on average. Finally, we can also compute the accuracy metric using the confusion matrix.
```

File: docs/JAX_examples_image_segmentation.md [hunk 4]
```diff
@@ -0,0 +1,1387 @@
...
+
+def compute_mean_iou(cm: jax.Array) -> jax.Array:
+    return compute_iou(cm).mean()
+
+
+def compute_accuracy(cm: jax.Array) -> jax.Array:
+    return jnp.diag(cm).sum() / (cm.sum() + 1e-15)
+```
+
+Let us now define training and evaluation steps
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1160]:
```suggestion
Next, let's define training and evaluation steps:
```

File: docs/JAX_examples_image_segmentation.md [hunk 5]
```diff
@@ -0,0 +1,1387 @@
...
+
+    eval_metrics.update(
+        total_loss=loss,
+        y_pred=logits,
+        y=masks,
+    )  # In-place updates.
+```
+
+Let us now define metrics we want to compute during the evaluation phase: total loss and confusion matrix computed on training and validation datasets. Finally, we define helper objects to store metrics history.
+Metrics like IoU per class, mean IoU and accuracy will be computed using the confusion matrix in the evaluation code.
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1197]:
```suggestion
We will also define metrics we want to compute during the evaluation phase: total loss and confusion matrix computed on training and validation datasets. Finally, we define helper objects to store the metrics history.
Metrics like IoU per class, mean IoU and accuracy will be computed using the confusion matrix in the evaluation code.
```

File: docs/JAX_examples_image_segmentation.md [hunk 6]
```diff
@@ -0,0 +1,1387 @@
...
+for epoch in range(num_epochs):
+    train_one_epoch(epoch)
+    if (epoch % 3 == 0) or (epoch == num_epochs - 1):
+        val_mean_iou = evaluate_model(epoch)
+        if val_mean_iou > best_val_mean_iou:
+            save_model(epoch)
+            best_val_mean_iou = val_mean_iou
+```
+
+Let us check saved models
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1330]:
```suggestion
We can check the saved models:
```

File: docs/JAX_examples_image_segmentation.md [hunk 7]
```diff
@@ -0,0 +1,1387 @@
...
+
+```{code-cell} ipython3
+!ls /tmp/output-oxford-model/
+```
+
+Let us visualize collected metrics
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1336]:
```suggestion
and visualize collected metrics:
```

File: docs/JAX_examples_image_segmentation.md [hunk 8]
```diff
@@ -0,0 +1,1387 @@
...
+plt.legend()
+```
+
+```{code-cell} ipython3
+plt.plot(eval_metrics_history["train_mean_IoU"], label="Mean IoU on training set")
+plt.plot(eval_metrics_history["val_mean_IoU"], label="Mean IoU on validation set")
+plt.legend()
+```
+
+Let us visualize model predictions on validation data
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1350]:
```suggestion
Next, we will visualize model predictions on validation data:
```

File: docs/JAX_examples_image_segmentation.md [hunk 9]
```diff
@@ -0,0 +1,1387 @@
...
+    axs[4].imshow(img)
+    axs[4].imshow(pred, alpha=0.5)
+```
+
+```{code-cell} ipython3
+for img, mask, pred in zip(images[:4], masks[:4], preds[:4]):
+    display_image_mask_pred(img, mask, pred, label=" (validation set)")
+```
+
+We can see that model can roughly predict the shape of the animal and the background and struggles with predicting the boundary. We can also observe that model starts overfitting according to the training and validation loss curves on the above plot. Carefully choosing hyperparameters and another optimizer like Adam we may achieve better results.
```
Comment (melissawm) [inline docs/JAX_examples_image_segmentation.md:1387]:
```suggestion
We can see that model can roughly predict the shape of the animal and the background and struggles with predicting the boundary. We can also observe that the model starts overfitting according to the training and validation loss curves on the above plot. Carefully choosing hyperparameters and another optimizer like Adam we may achieve better results.
```

**Review by melissawm — CHANGES_REQUESTED** [review]
Left some textual suggestions but mainly I can't get the notebook to run completely, potentially because of recent changes to flax/nnx.

File: docs/JAX_examples_image_segmentation.md [hunk 10]
```diff
@@ -0,0 +1,1387 @@
...
+This approach can be easily extended to any image segmentation datasets and users can reuse this code for their own datasets.
+
+In the code below we make a choice of using OpenCV and Pillow to read images and masks as NumPy arrays, [Albumentations](https://github.com/albumentations-team/albumentations) for data augmentations and
+[`grain`](https://github.com/google/grain/) for batched data loading. This choice of tools is not restricted and can be changed to any other alternative ones, e.g. [tensorflow_dataset](https://www.tensorflow.org/datasets) or [torchvision](https://pytorch.org/vision/stable/index.html).
+
++++
+
+### Requirements installation
+
+Let's install required python packages
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply docs/JAX_examples_image_segmentation.md:35]: It is described just above: 
> In the code below we make a choice of using OpenCV and Pillow to read images and masks as NumPy arrays, [Albumentations](https://github.com/albumentations-team/albumentations) for data augmentations and
[`grain`](https://github.com/google/grain/) for batched data loading. This choice of tools is not restricted and can be changed to any other alternative ones, e.g. [tensorflow_dataset](https://www.tensorflow.org/datasets) or [torchvision](https://pytorch.org/vision/stable/index.html).

File: docs/JAX_examples_image_segmentation.md [hunk 11]
```diff
@@ -0,0 +1,1387 @@
...
+
+        super().__init__(*layers)
+```
+
+```{code-cell} ipython3
+---
+jupyter:
+  source_hidden: true
+---
+class InstanceNorm(nnx.GroupNorm):
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply docs/JAX_examples_image_segmentation.md:616]: Yes, there is a typo in `PatchEmbeddingBlock`, thansk for catching it

File: docs/JAX_examples_image_segmentation.md [hunk 12]
```diff
@@ -0,0 +1,1387 @@
...
+
+```{code-cell} ipython3
+# nnx.display(model)
+```
+
+## Train the model
+
+In previous sections we defined training and validation dataloaders and the model. In this section we will train the model and define the loss function and the optimizer to perform the parameters optimization.
+
+For semantic segmentation task we can define the loss function as a sum of Cross-Entropy loss and Jaccard losses.
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply docs/JAX_examples_image_segmentation.md:1008]: I can add the following:
> The Cross-Entropy loss function is a standard loss function for a multi-class classification tasks and the Jaccard loss function helps directly optimizing Intersection-over-Union measure for semantic segmentation.

I do not have a good reference on Jaccard Loss function except this one : https://arxiv.org/pdf/1705.08790 where they proposed Lovasz-Softmax loss as another surrogate loss function to optimize IoU measure.

File: docs/JAX_examples_image_segmentation.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (melissawm) [reply docs/JAX_examples_image_segmentation.md:35]: OOps sorry I missed that! Feel free to ignore then :)

**Comment by jakevdp** [comment]
I'm not sure how to address that.

File: docs/JAX_examples_image_segmentation.md [hunk 13]
```diff
@@ -0,0 +1,1407 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.15.2
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Image segmentation with JAX
```
Comment (jakevdp) [inline docs/JAX_examples_image_segmentation.md:15]: Let's change this to something more specific, like "UNETR for image segmentation"

File: docs/JAX_examples_image_segmentation.md [hunk 14]
```diff
@@ -0,0 +1,1407 @@
...
+```{code-cell} ipython3
+for img, mask, pred in zip(images[:4], masks[:4], preds[:4]):
+    display_image_mask_pred(img, mask, pred, label=" (validation set)")
+```
+
+We can see that model can roughly predict the shape of the animal and the background and struggles with predicting the boundary. Carefully choosing hyperparameters we may achieve better results.
+
+```{code-cell} ipython3
+
+```
```
Comment (jakevdp) [inline docs/JAX_examples_image_segmentation.md:1407]: Remove empty cell

**Review by jakevdp — COMMENTED** [review]
Looks good overall – in the notebook, there are a lot of warnings about corrupt JPEG data. Should we worry about these? Is there any way to silence them?

**Comment by jakevdp** [comment]
> > there are a lot of warnings about corrupt JPEG data
> 
> There few corrupted images and warning is raised by OpenCV when reading them, in this case the fallback is using Pillow to read them. Using only Pillow for reading images is a bit slow. I can try to slience the warnings

Sounds good. Alternatively you could call these out and mention why they're non-problematic in this case.

**Comment by vfdev-5** [comment]
@jakevdp I applied the suggestions and removed indices of the data which raised warnings. Please review again. Thanks!

**Review by jakevdp — APPROVED** [review]
